src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1327,40 +1326,93 @@ protected override void AllocAndTransformNativeToManaged(ILCodeStream codeStream
+            if (ShouldBePinned)
+            {
+                //
+                // Pin the string and push a pointer to the first character on the stack.
+                //
+                TypeDesc stringType = Context.GetWellKnownType(WellKnownType.String);
 
-            ILLocalVariable vPinnedString = emitter.NewLocal(stringType, true);
-            ILCodeLabel lNullString = emitter.NewCodeLabel();
+                ILLocalVariable vPinnedString = emitter.NewLocal(stringType, true);
+                ILCodeLabel lNullString = emitter.NewCodeLabel();
 
-            LoadManagedValue(codeStream);
-            codeStream.EmitStLoc(vPinnedString);
-            codeStream.EmitLdLoc(vPinnedString);
+                LoadManagedValue(codeStream);
+                codeStream.EmitStLoc(vPinnedString);
+                codeStream.EmitLdLoc(vPinnedString);
 
-            codeStream.Emit(ILOpcode.conv_i);
-            codeStream.Emit(ILOpcode.dup);
+                codeStream.Emit(ILOpcode.conv_i);
+                codeStream.Emit(ILOpcode.dup);
 
-            // Marshalling a null string?
-            codeStream.Emit(ILOpcode.brfalse, lNullString);
+                // Marshalling a null string?
+                codeStream.Emit(ILOpcode.brfalse, lNullString);
 
-            codeStream.Emit(ILOpcode.call, emitter.NewToken(
-                Context.SystemModule.
-                    GetKnownType("System.Runtime.CompilerServices", "RuntimeHelpers").
-                        GetKnownMethod("get_OffsetToStringData", null)));
+                codeStream.Emit(ILOpcode.call, emitter.NewToken(
+                    Context.SystemModule.
+                        GetKnownType("System.Runtime.CompilerServices", "RuntimeHelpers").
+                            GetKnownMethod("get_OffsetToStringData", null)));
 
-            codeStream.Emit(ILOpcode.add);
+                codeStream.Emit(ILOpcode.add);
 
-            codeStream.EmitLabel(lNullString);
-            StoreNativeValue(codeStream);
+                codeStream.EmitLabel(lNullString);
+                StoreNativeValue(codeStream);
+            }
+            else
+            {
+                var helper = Context.GetHelperEntryPoint("InteropHelpers", "StringToUnicodeBuffer");
+                LoadManagedValue(codeStream);
+
+                codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
+
+                StoreNativeValue(codeStream);
+            }
         }
-    }
 
+        protected override void TransformNativeToManaged(ILCodeStream codeStream)
+        {
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            var helper = Context.GetHelperEntryPoint("InteropHelpers", "UnicodeBufferToString");
+            LoadNativeValue(codeStream);
+            codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
+            StoreManagedValue(codeStream);
+        }
+
+        protected override void EmitCleanupManagedToNative(ILCodeStream codeStream)
```

Comment:
You might want to change this to EmitCleanupManaged/EmitCleanupNative. It's easier to understand that way.